blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+                    if let Err(e) = handle.cancel_sender.send(()) {
+                        log::info!("Error sendin cancel job event: {e:?}");
+                    }
+                }
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn get_job_count(&self) -> Result<usize, JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if self.get_job_count_sender.send(result_sender).is_err() {
+            return Err(JobManagerError::FailedToSendGetJobCountEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => Ok(v),
+                Err(_) => Err(JobManagerError::FailedToReadJobCount),
+            }
+        }
+    }
+
+    fn get_job_count_handler(
+        jobs: &BTreeMap<JobKey, JobHandle>,
+        event: Option<oneshot::Sender<usize>>,
+    ) {
+        if let Some(result_sender) = event {
+            if let Err(e) = result_sender.send(jobs.len()) {
+                log::info!("Error sending get job count results: {e:?}");
+            }
+        }
+    }
+
+    pub async fn new_job(
+        &self,
+        block_id: Id<GenBlock>,
+    ) -> Result<(JobKey, UnboundedReceiver<()>), JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+        let (cancel_sender, cancel_receiver) = unbounded_channel::<()>();
+
+        if self.new_job_sender.send((block_id, cancel_sender, result_sender)).is_err() {
+            return Err(JobManagerError::FailedToSendNewJobEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => match v {
+                    Ok(v) => Ok((v, cancel_receiver)),
+                    Err(e) => Err(e),
+                },
+                Err(_) => Err(JobManagerError::FailedToCreateJob)
+            }
+        }
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn new_job_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<(
+            Id<GenBlock>,
+            UnboundedSender<()>,
+            oneshot::Sender<Result<JobKey, JobManagerError>>,
+        )>,
+    ) {
+        if let Some((current_tip_id, cancel_sender, result_sender)) = event {
+            let job_key = JobKey::new(current_tip_id);
+
+            if jobs.contains_key(&job_key) {
+                if let Err(e) = result_sender.send(Err(JobManagerError::JobAlreadyExists)) {
+                    log::info!("Error sending new job exists error: {e:?}");
+                }
+            } else {
+                jobs.insert(job_key.clone(), JobHandle { cancel_sender });
+
+                if let Err(e) = result_sender.send(Ok(job_key)) {
+                    log::info!("Error sending new job event: {e:?}");
+                }
+            }
+        }
+    }
+
+    pub async fn stop_job(&mut self, job_key: Option<JobKey>) -> Result<usize, JobManagerError> {
```

Comment:
Not a fan of the `Option` argument. The call `man.stop_job(None)` reads like you're not stopping any job.

I can see two possible ways to improve it:
* Either make this a private method (something like `stop_job_internal`) and have `stop_job` and `stop_jobs` wrappers, or
* Define a custom enum in place of `Option`, something like `enum JobSelector { Exact(JobKey), All }`

Reply:
Yep, you're right. I've opted for the first (but forgot to make it `private`...